File: haskell/src/Solutions/Euler183.hs
module Solutions.Euler183 where
-- Maximum is achieved at k closest to n/e.

uberdiv n k = if n `mod` k == 0 then uberdiv (n `div` k) k else n
smallest :: Integer -> Integer
smallest n =
    let k = round $ fromIntegral n / exp 1
    in if (k `div` gcd n k) `uberdiv` 2 `uberdiv` 5 == 1 then -n else n

euler183 :: IO String
euler183 = return $ show $ sum $ map smallest [5..10000]

File: rust/src/solutions/euler183.rs
// Maximum is achieved at k closest to n/e.

fn uberdiv(n: i64, k: i64) -> i64 {
    if n % k == 0 {
        uberdiv(n / k, k)
    } else {
        n
    }
}

fn smallest(n: i64) -> i64 {
    let k = ((n as f64) / std::f64::consts::E).round() as i64;
    if uberdiv(uberdiv(k / num::integer::gcd(n, k), 2), 5) == 1 {
        -n
    } else {
        n
    }
}

pub fn euler183() -> String {
    (5..=10000).map(smallest).sum::<i64>().to_string()
}
